### ISSUE TITLE
Containerize the security pipeline tools

### ISSUE THREAD

**Issue by bitkeks** [SovereignCloudStack/issues#526]
As an operator, I want to run the security pipeline tools in containers so that they can easily be deployed in a CI/CD pipeline like Zuul.

Belongs to epic #525.

## Definition of Ready:

- [x] User Story is small enough to be finished within one sprint
- [x] User Story is clear and understood by the whole team
- [ ] Acceptance criteria are defined
- [ ] Acceptance criteria are clear and understood by the whole team

## Definition of Done:

- [ ] All acceptance criteria are met
- [ ] Changes have been reviewed
- [ ] CI tests have run successfully
- [ ] Documentation has been updated
- [ ] Release Notes have been updated

**Comment by 90n20** [comment]
We have reviewed and tested the available official containers for the proposed tools.

- **ProjectDiscovery (Naabu, httpx, nuclei)** => [DockerHub site](https://hub.docker.com/u/projectdiscovery). Official containers are updated on each mayor release of the tools, ensuring that they are being run in their latest versions. They are fully functional and integrable into the proposed pipeline.

- **OWASP ZAP** => [Zap Docker](https://www.zaproxy.org/docs/docker/about/). There are 4 active official containers: stable, weekly, nightly and bare. We found that the most reliable ones are weekly, as it provides core and addon updates in a weekly basis (avoiding problems of everyday builds) and bare, as it is a small version of the stable one , ideal for CI environments, but lacking testing scripts.

- **Greenbone Community Edition** => [Containers](https://greenbone.github.io/docs/latest/22.4/container/index.html). Greenbone Community Edition does not run on a single container, in fact, it is a distributed service architecture, with each service running in a dedicated container. This could be a problem to integrate the tool in the pipeline due to its complexity and size, especially regarding its feed updates that are available on a daily basis. At this point, we are investigating three possible solutions:

    - Deploy the tool in the pipeline as it is, with all its services and containers. In this approach, a persistent volume will be needed in order to store the feeds and keep them on their latest versions (as they are updated daily, another pipeline could be used to maintain the feeds on the volume updated). This will add complexity and extra build time.
    - Use a non official single container image, like [Immauss](https://immauss.github.io/openvas/) (licensed as AGPL-3.0). It allows to perform a full scanner with only one container that includes all services and their requirements. The image itself contains a ready to use feed database and it is updated weekly.
    - Develop our own single containerized solution. This will involve research, development, integration and testing  in exchange of more control of what is being during the scans.
 
Additionally, we suggest to expand the proposed methodology and define two pipelines based on the tests done with the tools:

 - Daily (Naabu + Httpx + Nuclei + ZAP Baseline Scan): Fast scan.
 - Weekly, major commit or release (Naabu + Httpx + Nuclei + ZAP Full Scan + Greenbone CE): Slow and time intensive scan.

**Comment by bitkeks** [comment]
Regarding Greenbone / OpenVAS:

> At this point, we are investigating three possible solutions:

Please follow solution 1, deploy it as-is from Greenbone. 

Since we use full VMs, the creation of the GVM from Docker Compose should work. Then only the persistence of feed data is to be solved. 

Please also evaluate the re-use of VMs in this pipeline, maybe we can even skip the creation of the VM from scratch.

**Comment by 90n20** [comment]
Following the abovementioned solution and going deeper through the available Greenbone documentation on containers workflow, we have found that feeds are provided in their own container images, which are updated every 24 hours (as can be seen in Docker Hub logs). 

With this in mind, when the pentesting pipeline job is triggered and Greenbone container architecture is deployed, via its docker compose file, latest images versions are pulled, hence including feed updates for that day. 

However it still takes a bit of time for deployed services daemons to load the feeds into memory. As per our tests in a controlled environment, this time ranges from 15 to 30 minutes, time which we think is more than **acceptable** and should let us get rid of mantain persistent volumes or use extra pipelines to update the feeds.

The containers with feed updates are the following:
- notus-data
- vulnerability-tests
- scap-data
- dfn-cert-data
- cert-bund-data
- report-formats
- data-objects

**Comment by bitkeks** [comment]
Great work, thanks!

This looks acceptable to me as well. Let's continue with this approach and build the pipeline without persistent container storage. 

We'll see how it works out in SCS Zuul when deployed. The internal SCS registry can cache the feed containers as well.

**Comment by 90n20** [comment]
Tests are being performed wit ZAP, as containerized scans (both proposed Baseline and Full) only allow to perform a scan against a single host/target. We are investigating if creating a specific scanning plan through [ZAP automation framework](https://www.zaproxy.org/docs/desktop/addons/automation-framework) could handle this or if we should launch a dedicated containerized scan for each defined target.

Regarding Greenbone, we have observed that the best approach to fit our needs is to interact with _gvmd_ daemon using the [python-gvm](https://greenbone.github.io/python-gvm/) library, which allows to interact with the ospd-openvas scanner leveraging the Greenbone Management Protocol (using  _gvm-tools_ could be an alternative too).

**Comment by bitkeks** [comment]
> We are investigating if creating a specific scanning plan through [ZAP automation framework](https://www.zaproxy.org/docs/desktop/addons/automation-framework) could handle this or if we should launch a dedicated containerized scan for each defined target.

lgtm!

> is to interact with gvmd daemon using the [python-gvm](https://greenbone.github.io/python-gvm/) library

lgtm!

**Comment by 90n20** [comment]
DefectDojo instance has been configured in order to define and create "products", "engagements" and "tests", as well as API integrations as follows (subject to changes):

- Product Types: 
  - SCS => Placeholder for everything related to SCS environment
- Products: 
   - Testbed => At present our own deployed testbed is the target of the scans being performed
     - Engagements:
       - Daily Testbed Scan
       - Weekly Testbed Scan
   - Test MineryReport => Used for debug the integration through automated scans performed from our private Zuul instance
     - Engagements:
       - Test Daily Testbed Scan

Pipeline in our test environment  has been modified to export scan results to this DefectDojo instance. This will be commited to [security-infra-scan-pipeline](https://github.com/SovereignCloudStack/security-infra-scan-pipeline) repository as soon as we confirm that everything work as expected, removing scan results from pipeline output.

![image](https://github.com/SovereignCloudStack/issues/assets/54727263/78cc3212-7de8-41ab-b673-8c870552e731)

![image](https://github.com/SovereignCloudStack/issues/assets/54727263/de8ee61d-9175-4496-ab9f-7465bcd8c01d)

![image](https://github.com/SovereignCloudStack/issues/assets/54727263/b6bc799d-7c7c-4867-bec6-9d8dedd0cca0)

In other topic, DefectDojo user authentication can be performed with Keycloak (as OAuth2 provider) using the social-auth plugin. In order for this to work, a client of type _openid-connect_ should be created in the keycloak realm. It could be worth it to use the available SCS authentication system to allow users to login to the instance.

**Comment by bitkeks** [comment]
Great! Is the test instance publicly accessible?
Maybe we can use Github as login source?

**Comment by 90n20** [comment]
> Great! Is the test instance publicly accessible? 

Yes, as it is hosted in our assigned project space. However, as for now and testing purposes, it only has administrative access. Our final goal is to add to the repository both terraform and ansible scripts to allow everyone to deploy the same architecture.
> Maybe we can use Github as login source?

The documentation only mentions Github Enterprise, but we are investigating in parallel with other tasks whether it is possible to use Github as an OAuth2 mechanism.

**Comment by bitkeks** [comment]
David sent me some credentials for testing. We'll skip Github auth, not needed for the demo setup at the moment

**Comment by bitkeks** [comment]
Done in repo https://github.com/SovereignCloudStack/security-infra-scan-pipeline/
